llvm/lib/CodeGen/MachineVerifier.cpp

```diff
@@ -1274,6 +1274,24 @@ void MachineVerifier::verifyPreISelGenericInstruction(const MachineInstr *MI) {
         if (TypeSize::isKnownGT(MMO.getSize().getValue(),
                                 ValTy.getSizeInBytes()))
           report("load memory size cannot exceed result size", MI);
+
+        if (MMO.getRanges()) {
+          auto operandBits = [](const MDOperand &o) -> unsigned {
+            ConstantInt *i = mdconst::dyn_extract<ConstantInt>(o);
+            if (!i->isNegative())
+              return i->getValue().getActiveBits();
+            APInt reversed(i->getValue());
```

Comment:
You shouldn't need to consider anything about the value of the range, only the type

Reply:
This is something I still do not understand:

> The verifier should reject cases like this:
> ...
> !0 = !{i64 -4294967295, i64 4294967296}
> ...
> %3:_(<2 x s32>) = G_LOAD %0(p1) :: (volatile load (s64), align 4, !range !0, addrspace 1)

but if we get the information about the type, it is "i64" and there is a match between 2*32 == 64, no issues, you mentioned vector or scalar is not important, what would be the reason to reject this instruction?

Reply:
@arsenm could you please help me understand this piece, it looks like if I rely on type information alone, the cases that are mentioned in the description of this bug would be still legal. Thank you!

Reply:
thank you for the explanation, I will get back with updated code

Reply:
updated, please take a look